### PR TITLE
Split `test/index.js` into separate typescript test files

### PR DESCRIPTION
**Pull request by ryanio**
This PR:

* Splits `test/index.js` file into separate files and upgrades them to typescript
* Adds eslint to `lint:check` and `lint:fix`
* Adds eslint `import/order` rule for more consistent import ordering

### CODE REVIEW

**Comment by ryanio** [comment]
one note, I renamed `rc-1.1` for the reference contract version to `1.1-reference` to reduce confusion with `rc` typically meaning `release candidate` when used as a version pre/postfix.

File: test/utils/fixtures/index.ts [hunk 1]
```diff
@@ -722,7 +725,7 @@ export const seaportFixture = async (owner: Wallet) => {
         if (consideration.itemType === 1) {
           // ERC20
           // search for transfer
-          const transferLogs = (tokenEvents || [])
+          const transferLogs = (tokenEvents ?? [])
```
Comment (stuckinaboot) [inline test/utils/fixtures/index.ts:728]: Minor: I think you should share with contributors the difference between `||` and `??` as ppl seemed to often use `||` throughout the test files in places where `??` was likely intended (if they knew about it) for "using right side only if left side is null or undefined"... and I have a feeling ppl will keep doing this is if it's not brought up haha

File: test/utils/helpers.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    .slice(0, 10);
+
+export const simulateMatchOrders = async (
+  marketplaceContract: ethers.Contract,
+  orders: Order[],
+  fulfillments: Fulfillment[],
+  caller: ethers.Wallet,
+  value: ethers.BigNumberish
+) => {
+  return marketplaceContract
```
Comment (stuckinaboot) [inline test/utils/helpers.ts:28]: nit: can remove `=> {
return ...
}` and just do `=> marketplaceContract...`

(just a minor style thing, your call here)

File: test/revert.spec.ts [hunk 3]
```diff
@@ -0,0 +1,5907 @@
...
+        consideration,
+        0 // FULL_OPEN
+      );
+
+      await expect(
+        marketplaceContract
+          .connect(buyer)
+          .fulfillAvailableOrders(
+            [order],
+            [[[0, 0]]] as any,
```
Comment (stuckinaboot) [inline test/revert.spec.ts:5819]: nit: is there a more specific type we can use? Does `fulfillAvailableOrders` have typed args?

File: test/revert.spec.ts [hunk 4]
```diff
@@ -0,0 +1,5907 @@
...
+        0 // FULL_OPEN
+      );
+
+      await expect(
+        marketplaceContract
+          .connect(buyer)
+          .fulfillAvailableAdvancedOrders(
+            [order],
+            [],
+            [[[0, 0]]] as any,
```
Comment (stuckinaboot) [inline test/revert.spec.ts:5843]: Same question as here https://github.com/ProjectOpenSea/seaport/pull/534/files#r907492307

**Review by stuckinaboot — APPROVED** [review]
Great job! This will be enourmously helpful going forward and help developers enjoy advancing Seaport tests

File: test/advanced.spec.ts [hunk 5]
```diff
@@ -0,0 +1,4205 @@
...
+        return receipt;
+      });
+
+      orderStatus = await marketplaceContract.getOrderStatus(orderHash);
+
+      expect({ ...orderStatus }).to.deep.equal(
+        buildOrderStatus(true, false, toBN(1), prime2)
+      );
+
+      order.numerator = prime1 as any;
```
[the remark this answers is not in the document]
Comment (ryanio) [reply test/advanced.spec.ts:820]: Yes numerator/denominator is number but the issue here is in this test specifically is it is an overflow test so converting to number got a BigNumber overflow error, so passing in as a BigNumber (overriding the type warning with `any`) allowed the overflow error to come from the contract level.

File: test/advanced.spec.ts [hunk 6]
```diff
@@ -0,0 +1,4205 @@
...
+      method: "hardhat_reset",
+    });
+  });
+
+  before(async () => {
+    owner = new ethers.Wallet(randomHex(32), provider);
+
+    await faucet(owner.address, provider);
+
+    ({
```
[the remark this answers is not in the document]
Comment (ryanio) [reply test/advanced.spec.ts:82]: It's a good idea, I was thinking about that as I was going but was unsure if it was set up the way it was for a reason. I will give it a try to convert it and it should really simplify the setup as you mention :)

File: test/utils/fixtures/index.ts (hunk 1 above)
Comment (ryanio) [reply test/utils/fixtures/index.ts:728]: Hehe yeah the nullish coalescing operator is really great. Where would you suggest I add this communication? I could add this lint rule actually to help enforce: https://github.com/typescript-eslint/typescript-eslint/blob/main/packages/eslint-plugin/docs/rules/prefer-nullish-coalescing.md

**Review by 0age — APPROVED** [review]
Great work here! Merge at wil

File: test/advanced.spec.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryanio) [reply test/advanced.spec.ts:820]: i'll add a comment to help clarify here

File: test/revert.spec.ts (hunk 3 above)
Comment (ryanio) [reply test/revert.spec.ts:5819]: it does, but it's expecting this generated type from `typechain-types`:
```typescript
export type FulfillmentComponentStruct = {
  orderIndex: BigNumberish;
  itemIndex: BigNumberish;
};
```
any thoughts how to handle so I can remove the `as any` ?

**Pull request merged**